<a href="https://colab.research.google.com/github/VictorPortolani/am-turma5/blob/main/Atividade_aula0908.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pipeline de Análise de Dados [Census]


## Importação de Bibliotecas

* Pandas
* Numpy
* MatPlot
* Seaborn
* Sklearn

In [219]:
!pip install pandas
!pip install numpy
!pip install matplotlib
!pip install seaborn
!pip install scikit-learn

In [160]:
import pandas as pd # organizar arrays
import numpy as np #funções matemáticas complexos
import matplotlib.pyplot as plt #plotar funções matemáticas complexas
import seaborn as sns # vizualizar gráficos
import plotly.express as px #expor a biblioteca

In [161]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [162]:
base_census = pd.read_csv('/content/drive/MyDrive/am-t5/census.csv', skipinitialspace=True)


## Carregamento dos Dados da Base

In [163]:
base_census

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [164]:
base_census.describe()
#income: renda
#loan: divida
#defalut: sim ou não para dívida
#std: desvio padrão
#mean: média

,age,final-weight,education-num,capital-gain,capital-loos,hour-per-week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


## Características da Base
- População  amostral: 32561
- Minorias Importantes: idade = 28 e 90, ed. = 9 e 16, ht trabalhadas = 40 e 99
- Media amostral: idade = 38.58, ed. = 10, ganho = 1077.65, perca = 87.30, ht = 40.44
- Desvio padrão [limites]: idade = 24.94, ed. = 7.51, ganho = -6757.64,
perda = -315.66, ht = 28.9   

In [211]:
grafic_class_wa = px.treemap(base_census, path=['workclass', 'age'])
grafic_class_wa.show()

In [212]:
grafic_class_ora = px.treemap(base_census, path=['occupation', 'relationship', 'age'])
grafic_class_ora.show()

In [213]:
grafic_finetune_oeg = px.parallel_categories(base_census, dimensions=['education', 'race', 'income'])
grafic_finetune_oeg.show()

In [214]:
grafic_finetune_owg = px.parallel_categories(base_census, dimensions=['workclass', 'race', 'occupation'])
grafic_finetune_owg.show()

##Análise de Anomalias de Dados [Census]
**Enunciado: Auditoria e Higienização de Dados Censitários**

Vamos analizar e estruturar a ingestão de um novo pipeline de dados sociodemográficos focado no desenvolvimento de modelos de inteligência artificial. O arquivo **census.csv** foi extraído de um sistema legado, mas o processo automatizado falhou antes mesmo da etapa de modelagem estatística. Como futuros cientistas da computação e engenheiros de dados, a missão de vocês é realizar uma auditoria rigorosa neste arquivo para garantir a integridade da informação.

Dicas de Investigação e Solução

* **Identificação de Linhas Infiltradas**: Ao forçarem o carregamento inicial (pulando as linhas defeituosas do topo), observem o comportamento dos tipos de dados informados pelo `.dtypes`. Se colunas que deveriam ser numéricas (`age` ou `hour-per-week`) foram lidas como `object`, existem *strings* perdidas no meio da massa de dados. O cabeçalho e fragmentos de texto corrompido estão repetidos ao longo do arquivo.
* **Inconsistência Categórica e Caracteres de Escape:** Analisem a distribuição de classes na coluna `income` através do `.value_counts()`. A mesma classe de renda foi acidentalmente dividida em múltiplas categorias distintas devido a falhas de exportação com caracteres de escape textuais (como `\<=50K` versus `<=50K`).
* **Dados Faltantes Não Convencionais:** Avaliem como a ausência de informação foi mascarada pelo sistema de origem. Em bases censitárias clássicas, campos em branco (como `workclass` e `native-country`) frequentemente não aparecem como `NaN` estruturados, mas são substituídos por marcadores categóricos arbitrários, como o caractere `?`.
* **Tratamento de Espaçamento:** A separação por vírgulas pode conter espaços em branco residuais no início das palavras, prejudicando filtros e agrupamentos futuros. O parâmetro `skipinitialspace=True` na leitura do Pandas é um atalho eficiente para padronizar as bordas do texto.

In [169]:
base_census.dtypes
# não há objetos em valores que deveriam ser numéricos

,0
age,int64
workclass,object
final-weight,int64
education,object
education-num,int64
marital-status,object
occupation,object
relationship,object
race,object
sex,object


In [170]:
base_census.loc[base_census.isnull().any(axis=1)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


In [171]:
colunas_texto = base_census.select_dtypes(include=['object'])

# Aplica o .strip() para limpar espaços e depois compara com "?"
(colunas_texto.apply(lambda x: x.str.strip()) == "?").sum()

,0
workclass,1836
education,0
marital-status,0
occupation,1843
relationship,0
race,0
sex,0
native-country,583
income,0


In [172]:
# Filtra as linhas onde o valor " ?" aparece em pelo menos uma das colunas
base_census.loc[(base_census == ' ?').any(axis=1)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


In [173]:
base_census.loc[base_census['workclass'].str.contains(' ?', regex=False, na=False)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


In [174]:
base_census.loc[((base_census == ' ?').any(axis=1)) & (base_census['age'] < 24)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


Normalização de Estudantes para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Dependent` e `Student`.

In [175]:
def  normalizarEstudantes(df):
  condition_age_24 = df['age'] < 24
  df.loc[condition_age_24 & (df['workclass'] == '?'), 'workclass'] = 'Dependent'
  df.loc[condition_age_24 & (df['occupation'] == '?'), 'occupation'] = 'Student'
  return df
base_census = normalizarEstudantes(base_census)

In [176]:
base_census.loc[base_census['age'] < 18]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
106,17,Dependent,304873,10th,6,Never-married,Student,Own-child,White,Female,34095,0,32,United-States,<=50K
209,17,Private,65368,11th,7,Never-married,Sales,Own-child,White,Female,0,0,12,United-States,<=50K
262,17,Private,245918,11th,7,Never-married,Other-service,Own-child,White,Male,0,0,12,United-States,<=50K
271,17,Private,191260,9th,5,Never-married,Other-service,Own-child,White,Male,1055,0,24,United-States,<=50K
335,17,Private,270942,5th-6th,3,Never-married,Other-service,Other-relative,White,Male,0,0,48,Mexico,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31772,17,Dependent,256173,10th,6,Never-married,Student,Own-child,White,Female,0,0,15,United-States,<=50K
31864,17,Private,214787,12th,8,Never-married,Adm-clerical,Own-child,White,Female,0,0,25,United-States,<=50K
31959,17,Private,24090,HS-grad,9,Never-married,Exec-managerial,Own-child,White,Female,0,0,35,United-States,<=50K
32282,17,Private,311907,11th,7,Never-married,Other-service,Own-child,White,Male,0,0,25,United-States,<=50K


In [177]:
base_census.loc[((base_census == '?').any(axis=1)) & (base_census['age'] >= 62)]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
77,67,?,212759,10th,6,Married-civ-spouse,?,Husband,White,Male,0,0,2,United-States,<=50K
160,68,?,38317,1st-4th,2,Divorced,?,Not-in-family,White,Female,0,0,20,United-States,<=50K
221,64,?,187656,1st-4th,2,Divorced,?,Not-in-family,White,Male,0,0,40,United-States,<=50K
346,67,?,36135,11th,7,Married-civ-spouse,?,Husband,White,Male,0,0,8,United-States,<=50K
430,80,?,107762,HS-grad,9,Widowed,?,Not-in-family,White,Male,0,0,24,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32343,66,?,93318,HS-grad,9,Widowed,?,Unmarried,White,Female,0,0,40,United-States,<=50K
32494,82,?,403910,HS-grad,9,Never-married,?,Not-in-family,White,Male,0,0,3,United-States,<=50K
32525,81,?,120478,Assoc-voc,11,Divorced,?,Unmarried,White,Female,0,0,1,?,<=50K
32539,71,?,287372,Doctorate,16,Married-civ-spouse,?,Husband,White,Male,0,0,10,United-States,>50K


- Normalização de Idosos para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Social Security` e `Retired`.
- Normalização de Desempregados para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Unemployed` e `No Market`.
- Normalização de Trabalhadores Independentes para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Independent` e `Family Support`.
- Normalização de Investidores para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Financial Market` e `Investors`.

In [178]:
base_census[((base_census['age'] >=24) & (base_census['age'] <=62)) & ((base_census['workclass'] == '?') & (base_census['capital-gain'] == 0))]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
27,54,?,180211,Some-college,10,Married-civ-spouse,?,Husband,Asian-Pac-Islander,Male,0,0,60,South,>50K
61,32,?,293936,7th-8th,4,Married-spouse-absent,?,Not-in-family,White,Male,0,0,40,?,<=50K
69,25,?,200681,Some-college,10,Never-married,?,Own-child,White,Male,0,0,40,United-States,<=50K
128,35,?,129305,HS-grad,9,Married-civ-spouse,?,Husband,White,Male,0,0,40,United-States,<=50K
149,43,?,174662,Some-college,10,Divorced,?,Not-in-family,White,Female,0,0,40,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32477,49,?,114648,12th,8,Divorced,?,Other-relative,Black,Male,0,0,40,United-States,<=50K
32490,60,?,134152,9th,5,Divorced,?,Not-in-family,Black,Male,0,0,35,United-States,<=50K
32530,35,?,320084,Bachelors,13,Married-civ-spouse,?,Wife,White,Female,0,0,55,United-States,>50K
32531,30,?,33811,Bachelors,13,Never-married,?,Not-in-family,Asian-Pac-Islander,Female,0,0,99,United-States,<=50K


In [179]:
def normalizarCenso(df):
    condicao_aposentado = df['age'] > 62
    df.loc[condicao_aposentado & (df['workclass'] == '?'), 'workclass'] = 'Social Security'
    df.loc[condicao_aposentado & (df['occupation'] == '?'), 'occupation'] = 'Retired'

    condicao_desempregado = (df['age'] >=24) & (df['age'] <=62) & (df['capital-gain'] == 0) & (df['hour-per-week'] < 1)
    df.loc[condicao_desempregado & (df['workclass'] == '?'), 'workclass'] = 'Unemployed'
    df.loc[condicao_desempregado & (df['occupation'] == '?'), 'occupation'] = 'No Market'

    return df

base_census = normalizarCenso(base_census)

In [180]:
base_census.loc[base_census['age'] >= 62]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
74,79,Private,124744,Some-college,10,Married-civ-spouse,Prof-specialty,Other-relative,White,Male,0,0,20,United-States,<=50K
77,67,Social Security,212759,10th,6,Married-civ-spouse,Retired,Husband,White,Male,0,0,2,United-States,<=50K
100,76,Private,124191,Masters,14,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,40,United-States,>50K
144,70,Private,105376,Some-college,10,Never-married,Tech-support,Other-relative,White,Male,0,0,40,United-States,<=50K
148,64,Private,181232,11th,7,Married-civ-spouse,Craft-repair,Husband,White,Male,0,2179,40,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32515,66,Federal-gov,47358,10th,6,Married-civ-spouse,Craft-repair,Husband,White,Male,3471,0,40,United-States,<=50K
32525,81,Social Security,120478,Assoc-voc,11,Divorced,Retired,Unmarried,White,Female,0,0,1,?,<=50K
32539,71,Social Security,287372,Doctorate,16,Married-civ-spouse,Retired,Husband,White,Male,0,0,10,United-States,>50K
32542,72,Social Security,129912,HS-grad,9,Married-civ-spouse,Retired,Husband,White,Male,0,0,25,United-States,<=50K


In [191]:
base_census[((base_census['age'] >=24) & (base_census['age'] <=62) & (base_census['hour-per-week'] > 0) & (base_census['workclass'] == '?'))]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
27,54,?,180211,Some-college,10,Married-civ-spouse,?,Husband,Asian-Pac-Islander,Male,0,0,60,South,>50K
61,32,?,293936,7th-8th,4,Married-spouse-absent,?,Not-in-family,White,Male,0,0,40,?,<=50K
69,25,?,200681,Some-college,10,Never-married,?,Own-child,White,Male,0,0,40,United-States,<=50K
128,35,?,129305,HS-grad,9,Married-civ-spouse,?,Husband,White,Male,0,0,40,United-States,<=50K
149,43,?,174662,Some-college,10,Divorced,?,Not-in-family,White,Female,0,0,40,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32477,49,?,114648,12th,8,Divorced,?,Other-relative,Black,Male,0,0,40,United-States,<=50K
32490,60,?,134152,9th,5,Divorced,?,Not-in-family,Black,Male,0,0,35,United-States,<=50K
32530,35,?,320084,Bachelors,13,Married-civ-spouse,?,Wife,White,Female,0,0,55,United-States,>50K
32531,30,?,33811,Bachelors,13,Never-married,?,Not-in-family,Asian-Pac-Islander,Female,0,0,99,United-States,<=50K


In [192]:
def normalizarIndependetWorker(df):
    condition_Independent = (df['age'] >=24) & (df['age'] <=62) & (df['capital-gain'] == 0) & (df['hour-per-week'] > 0)
    df.loc[condition_Independent & (df['workclass'] == '?'), 'workclass'] = 'Independent'
    df.loc[condition_Independent & (df['occupation'] == '?'), 'occupation'] = 'Family Support'

    return df

base_census = normalizarIndependetWorker(base_census)

In [193]:
base_census[((base_census['age'] >=24) & (base_census['age'] <=62) & (base_census['hour-per-week'] > 0) & (base_census['workclass'] == '?'))]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
297,39,?,157443,Masters,14,Married-civ-spouse,?,Wife,Asian-Pac-Islander,Female,3464,0,40,?,<=50K
484,52,?,271749,12th,8,Never-married,?,Other-relative,Black,Male,594,0,40,United-States,<=50K
1349,62,?,225652,7th-8th,4,Married-civ-spouse,?,Husband,White,Male,3411,0,50,United-States,<=50K
1574,49,?,261059,10th,6,Separated,?,Own-child,White,Male,2176,0,40,United-States,<=50K
2513,47,?,174525,HS-grad,9,Married-civ-spouse,?,Husband,White,Male,3942,0,40,?,<=50K
3594,59,?,154236,HS-grad,9,Married-civ-spouse,?,Wife,White,Female,7688,0,40,United-States,>50K
3736,36,?,53606,Assoc-voc,11,Married-civ-spouse,?,Wife,White,Female,3908,0,8,United-States,<=50K
3863,62,?,94931,Assoc-voc,11,Married-civ-spouse,?,Husband,White,Male,3411,0,40,United-States,<=50K
4215,53,?,158352,Masters,14,Never-married,?,Not-in-family,White,Female,8614,0,35,United-States,>50K
4500,35,?,296738,11th,7,Separated,?,Not-in-family,White,Female,6849,0,60,United-States,<=50K


In [194]:
def normalizarInvestidor(df):
    condition_Independent = (df['age'] >=24) & (df['age'] <=62) & (df['capital-gain'] > 0) & (df['hour-per-week'] > 0)
    df.loc[condition_Independent & (df['workclass'] == '?'), 'workclass'] = 'Fiancial Market'
    df.loc[condition_Independent & (df['occupation'] == '?'), 'occupation'] = 'Investors'

    return df

base_census = normalizarInvestidor(base_census)

In [195]:
base_census.loc[((base_census == '?').any(axis=1))]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income
14,40,Private,121772,Assoc-voc,11,Married-civ-spouse,Craft-repair,Husband,Asian-Pac-Islander,Male,0,0,40,?,>50K
38,31,Private,84154,Some-college,10,Married-civ-spouse,Sales,Husband,White,Male,0,0,38,?,>50K
51,18,Private,226956,HS-grad,9,Never-married,Other-service,Own-child,White,Female,0,0,30,?,<=50K
61,32,Independent,293936,7th-8th,4,Married-spouse-absent,Family Support,Not-in-family,White,Male,0,0,40,?,<=50K
93,30,Private,117747,HS-grad,9,Married-civ-spouse,Sales,Wife,Asian-Pac-Islander,Female,0,1573,35,?,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32449,44,Self-emp-inc,71556,Masters,14,Married-civ-spouse,Sales,Husband,White,Male,0,0,50,?,>50K
32469,58,Self-emp-inc,181974,Doctorate,16,Never-married,Prof-specialty,Not-in-family,White,Female,0,0,99,?,<=50K
32492,42,Self-emp-not-inc,217597,HS-grad,9,Divorced,Sales,Own-child,White,Male,0,0,50,?,<=50K
32510,39,Private,107302,HS-grad,9,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,45,?,>50K


### Tratamento de `native-country`

In [209]:
def normalizarStatelessPersons(df):
    df.loc[(df['native-country'] == '?'), 'native-country'] = 'Stateless Persons'

    return df

base_census = normalizarStatelessPersons(base_census)

In [210]:
base_census.loc[((base_census == '?').any(axis=1))]

,age,workclass,final-weight,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loos,hour-per-week,native-country,income


### Resutado final
Normalização das Anomalias:
- Normalização de Idosos para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Social Security` e `Retired`.
- Normalização de Desempregados para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Unemployed` e `No Market`.
- Normalização de Trabalhadores Independentes para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Independent` e `Family Support`.
- Normalização de Investidores para `workclass` e `ocupation` assumindo os valores que estavam com `?` para valores `Financial Market` e `Investors`.
- A coluna `income` for feita a verificação e não valores declarados para múltiplas categorias.
- Para o `native-country` com `?` foi aplicado o valor `Stateless Persons`

# Pré - Processamento Escalonar [Censos]

In [222]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder, LabelEncoder
scale_census = StandardScaler()

In [223]:
X_census = base_census.iloc[:, 0:14].values

In [224]:
X_census

array([[39, 'State-gov', 77516, ..., 0, 40, 'United-States'],
       [50, 'Self-emp-not-inc', 83311, ..., 0, 13, 'United-States'],
       [38, 'Private', 215646, ..., 0, 40, 'United-States'],
       ...,
       [58, 'Private', 151910, ..., 0, 40, 'United-States'],
       [22, 'Private', 201490, ..., 0, 20, 'United-States'],
       [52, 'Self-emp-inc', 287927, ..., 0, 40, 'United-States']],
      dtype=object)

In [225]:
Y_census = base_census.iloc[:, 14].values

In [226]:
Y_census

array(['<=50K', '<=50K', '<=50K', ..., '<=50K', '<=50K', '>50K'],
      dtype=object)

In [230]:
(X_census[:, 2].min() - X_census[:, 2].max())

-1472420

In [231]:
X_census[:, 10].min() - X_census[:, 10].max()

-99999